src/rpc/index.js

```
@@ -41,63 +42,43 @@ module.exports = (dht) => {
+   * @param {string} props.protocol
+   * @param {DuplexStream} props.stream
+   * @param {Connection} props.connection connection
+   * @returns {Promise<void>}
    */
-  return function protocolHandler (protocol, conn) {
-    conn.getPeerInfo((err, peer) => {
-      if (err) {
-        log.error('Failed to get peer info')
-        log.error(err)
-        return
-      }
+  return async function onIncomingStream ({ protocol, stream, connection }) {
+    const peerInfo = await PeerInfo.create(connection.remotePeer)
+    peerInfo.protocols.add(protocol)
 
-      log('from: %s', peer.id.toB58String())
+    try {
+      await dht._add(peerInfo)
+    } catch (err) {
+      log.error(err)
+    }
 
-      pull(
-        conn,
-        lp.decode(),
-        pull.filter((msg) => msg.length < c.maxMessageSize),
-        pull.map((rawMsg) => {
-          let msg
-          try {
-            msg = Message.deserialize(rawMsg)
-          } catch (err) {
-            log.error('failed to read incoming message', err)
-            return
-          }
+    const idB58Str = peerInfo.id.toB58String()
+    log('from: %s', idB58Str)
 
-          return msg
-        }),
-        pull.filter(Boolean),
-        pull.asyncMap(async (msg, cb) => {
-          let response
-          try {
-            response = await handleMessage(peer, msg)
-          } catch (err) {
-            cb(err)
-          }
-          cb(null, response)
-        }),
-        // Not all handlers will return a response
-        pull.filter(Boolean),
-        pull.map((response) => {
-          let msg
-          try {
-            msg = response.serialize()
-          } catch (err) {
-            log.error('failed to send message', err)
-            return
-          }
-          return msg
-        }),
-        pull.filter(Boolean),
-        lp.encode(),
-        conn
-      )
-    })
+    await pipe(
+      stream.source,
+      lp.decode(),
+      utils.itFilter(
+        (msg) => msg.length < c.maxMessageSize
```

Same as above - we should consider doing more than failing silently